docs-internal/db.md

```diff
@@ -0,0 +1,364 @@
+
+## Example
+
+### Application Configuration File
+A configuration file used when creating a `Config` instance:
+```yaml
+helidon-db:
+    source: jdbc
+    url: jdbc:mysql://127.0.0.1:3306/myDatabase
+    username: user
+    password: password
+    statements:
+        insert-indexed-params: "INSERT INTO my_table VALUES(?, ?)"
+        insert-named-params:   "INSERT INTO my_table VALUES(:name, :type)"
+        select-all: "SELECT * FROM my_table"
+        select-indexed-params: "SELECT * FROM my_table WHERE name = ?"
+        select-named-params:   "SELECT * FROM my_table WHERE name = :name"
+```
+
+### Database Initialization:
+```java
+HelidonDb db = config
+                .get("helidon-db")
+                .as(HelidonDb::create)
+                .orElseThrow(() -> new IllegalStateException("Configuration is missing"));
+```
+
+### Statements Execution:
+
+#### Insert with Indexed Parameters
+```java
+db.execute(exec -> exec
+                .createNamedDmlStatement("insert-indexed-params")
+                .addParam("Pikachu")
+                .addParam("electric")
+                .execute())
+          .thenAccept(count -> <do something>)
+          .exceptionally(throwable -> <fail somehow>);
+```
+
+#### Insert with Named Parameters
+```java
+db.execute(exec -> exec
+                .createNamedDmlStatement("insert-named-params")
+                .addParam("name", "Pikachu")
+                .addParam("type", "electric")
+                .execute()
+          )
+          .thenAccept(count -> <do something>)
+          .exceptionally(throwable -> <fail somehow>);
+```
+
+#### Query without parameters
+
+```java
+db.execute(exec -> exec.namedQuery("select-all"))
+                .consume(response::send)
+                .exceptionally(throwable -> sendError(throwable, response));
+```
+
+#### Get with indexed parameters
+The `get` method is a shortcut to a query method that expects zero to 1 results.
+For Java 8, you need to use `OptionalHelper.from(maybeRow)` to be able to use `ifPresentOrElse`.
+When using newer versions of Java, `ifPresentOrElse` is available on `Optional`.
+```java
+db.execute(exec -> exec.namedGet("select-indexed-params"), "Pikachu")  
+          .thenAccept(maybeRow -> maybeRow.ifPresentOrElse(row -> <process row>,
+                                                            () -> <process not found>))
+          .exceptionally(throwable -> <fail somehow>);
+```
+
+#### Simple Transaction
+```java
+db.inTransaction(exec -> exec
+        .get("SELECT type FROM my_table WHERE name = ?", "Pikachu")
+        .thenAccept(maybeRow -> maybeRow
+                .ifPresent(row -> exec
+                        .insert("INSERT INTO my_table VALUES(?, ?)", "Raichu", row.column("type"))
+                )
+        )
+).exceptionally(throwable -> <fail somehow>)
+```
+
+### Interceptor implementation
+
+Example interceptor (for tracing):
+```java
+public class DbTracing implements DbInterceptor {
```

Comment:
How are these interceptors registered? Are they global or scoped to a certain type of statement or even an individual statement?

Reply:
Current when creating the `HelidonDb` instance - there is a builder method to add such an interceptor. So it is for the whole db.

Reply:
I have now modified the interceptor implementation
 - you can configure interceptors for
    - database
    - statement type (e.g. query, get, update, insert, delete, DML, unknown)
    - statement name (specific statements)

I have also added support for service loader using `DbInterceptorProvider`